bundles/org.openhab.binding.bluetooth/src/main/java/org/openhab/binding/bluetooth/discovery/internal/BluetoothDiscoveryProcess.java

```diff
@@ -0,0 +1,315 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link BluetoothDiscoveryProcess} does the work of creating a DiscoveryResult from a set of
+ * {@link BluetoothDisocveryParticipant}s
+ *
+ * @author Connor Petty - Initial Contribution
+ */
+@NonNullByDefault
+public class BluetoothDiscoveryProcess implements Supplier<DiscoveryResult>, BluetoothDeviceListener {
+
+    private static final int DISCOVERY_TTL = 300;
+
+    private final Logger logger = LoggerFactory.getLogger(BluetoothDiscoveryProcess.class);
+
+    private final Lock serviceDiscoveryLock = new ReentrantLock();
+    private final Condition connectionCondition = serviceDiscoveryLock.newCondition();
+    private final Condition serviceDiscoveryCondition = serviceDiscoveryLock.newCondition();
+    private final Condition nameDiscoveryCondition = serviceDiscoveryLock.newCondition();
+
+    private final Collection<BluetoothDiscoveryParticipant> participants;
+    private final BluetoothDevice device;
+
+    private volatile boolean servicesDiscovered = false;
+
+    public BluetoothDiscoveryProcess(BluetoothDevice device, Collection<BluetoothDiscoveryParticipant> participants) {
+        this.participants = participants;
+        this.device = device;
+    }
+
+    @Override
+    public DiscoveryResult get() {
+        // first see if any of the participants that don't require a connection recognize this device
+        List<BluetoothDiscoveryParticipant> connectionParticipants = new ArrayList<>();
+        for (BluetoothDiscoveryParticipant participant : participants) {
+            if (participant.requiresConnection(device)) {
+                connectionParticipants.add(participant);
+                continue;
+            }
+            try {
+                DiscoveryResult result = participant.createResult(device);
+                if (result != null) {
+                    return result;
+                }
+            } catch (RuntimeException e) {
+                logger.warn("Participant '{}' threw an exception", participant.getClass().getName(), e);
+            }
+        }
+
+        // Since we couldn't find a result, lets try the connection based participants
+        DiscoveryResult result = null;
+        if (!connectionParticipants.isEmpty()) {
+            BluetoothAddress address = device.getAddress();
+            try {
+                BluetoothAddressLocker.lock(address);
+                if (!device.hasListeners()) {
+                    result = findConnectionResult(connectionParticipants);
+                    // make sure to disconnect before letting go of the device
+                    if (device.getConnectionState() == ConnectionState.CONNECTED) {
+                        try {
+                            if (!device.disconnect()) {
+                                logger.debug("Failed to disconnect from device {}", address);
+                            }
+                        } catch (RuntimeException ex) {
```

Comment:
see above

Reply:
The underlying BluetoothDevice implementations are not flawless (especially the BluezBluetoothDevice), and it is very possible an unintended RuntimeException could occur here. But I don't think a failure to disconnect should prevent a device discovery hence we catch any RuntimeExceptions.